File: backend/app/limiters/leaky_bucket.py
from __future__ import annotations

import math

from redis.asyncio import Redis

from app.limiters.base import Decision


class LeakyBucketLimiter:
    algorithm = "leaky_bucket"

    async def allow(
        self,
        redis_client: Redis,
        policy_id: str,
        client_id: str,
        now_ms: int,
        params: dict[str, object],
    ) -> Decision:
        capacity = float(params["capacity"])
        leak_rate_per_sec = float(params["leak_rate_per_sec"])
        water_per_request = float(params.get("water_per_request", 1))

        key = f"rl:leaky:{policy_id}:{client_id}"
        water_raw, last_leak_raw = await redis_client.hmget(key, "water_level", "last_leak_ms")

        if water_raw is None or last_leak_raw is None:
            water_level = 0.0
            last_leak_ms = now_ms
        else:
            water_level = float(water_raw)
            last_leak_ms = int(float(last_leak_raw))

        elapsed_sec = max((now_ms - last_leak_ms) / 1000.0, 0.0)
        leaked_water = elapsed_sec * leak_rate_per_sec
        water_level = max(water_level - leaked_water, 0.0)

        if water_level + water_per_request <= capacity:
            water_level += water_per_request
            allowed = True
            reason = None
            retry_after_ms = None
        else:
            allowed = False
            reason = "rate_limit_exceeded"
            overflow = (water_level + water_per_request) - capacity
            retry_after_ms = max(int(math.ceil((overflow / leak_rate_per_sec) * 1000)), 1)

        ttl_sec = max(int(math.ceil(capacity / leak_rate_per_sec)) + 2, 2)
        await redis_client.hset(
            key,
            mapping={
                "water_level": f"{water_level:.8f}",
                "last_leak_ms": now_ms,
            },
        )
        await redis_client.expire(key, ttl_sec)

        remaining = int(max(math.floor((capacity - water_level) / water_per_request), 0))
        return Decision(
            allowed=allowed,
            reason=reason,
            remaining=remaining,
            retry_after_ms=retry_after_ms,
            algorithm_state={"water_level": round(water_level, 4)},
        )

File: gateway/internal/limiter/leaky_bucket.go
package limiter

import (
	"context"
	"fmt"
	"math"
)

type LeakyBucketLimiter struct {
	store RuntimeCounter
}

func NewLeakyBucketLimiter(store RuntimeCounter) *LeakyBucketLimiter {
	return &LeakyBucketLimiter{store: store}
}

func (l *LeakyBucketLimiter) Allow(ctx context.Context, policyID, clientID string, nowMS int64, params map[string]any) (Decision, error) {
	capacity, err := readPositiveInt(params, "capacity")
	if err != nil {
		return Decision{}, err
	}
	leakRate, err := readPositiveFloat(params, "leak_rate_per_sec")
	if err != nil {
		return Decision{}, err
	}
	waterPerRequest := 1
	if value, ok := params["water_per_request"]; ok {
		waterPerRequest, err = readPositiveInt(map[string]any{"water_per_request": value}, "water_per_request")
		if err != nil {
			return Decision{}, err
		}
	}

	key := fmt.Sprintf("rl:leaky_bucket:%s:%s", policyID, clientID)
	ttlSec := int(math.Ceil(float64(capacity)/leakRate)) + 1
	if ttlSec < 2 {
		ttlSec = 2
	}

	result, err := l.store.EvalLeakyBucket(ctx, key, nowMS, capacity, leakRate, waterPerRequest, ttlSec)
	if err != nil {
		return Decision{}, err
	}

	headroom := float64(capacity) - result.WaterLevel
	if headroom < 0 {
		headroom = 0
	}
	remaining := int(math.Floor(headroom / float64(waterPerRequest)))
	if remaining < 0 {
		remaining = 0
	}

	state := map[string]any{
		"count":                remaining,
		"window_start_ms":      nowMS,
		"window_size_sec":      0,
		"state_schema_version": 1,
		"water_level":          result.WaterLevel,
		"capacity":             capacity,
		"leak_rate_per_sec":    leakRate,
		"water_per_request":    waterPerRequest,
		"last_leak_ms":         result.LastLeakMS,
	}

	if result.Allowed {
		return Decision{
			Allowed:        true,
			Remaining:      &remaining,
			AlgorithmState: state,
		}, nil
	}

	reason := "rate_limit_exceeded"
	zero := 0
	retryAfter := result.RetryAfterMS
	return Decision{
		Allowed:        false,
		Reason:         &reason,
		Remaining:      &zero,
		RetryAfterMS:   &retryAfter,
		AlgorithmState: state,
	}, nil
}
